core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java

```diff
@@ -0,0 +1,165 @@
+     * </p>
+     * */
+    private final Pattern pattern;
+
+    /**
+     * Skeletal form of given path, which is used for duplicated routing rule detection.
+     * For example, "/{a}/{b}" and "/{c}/{d}" has same skeletal form and regarded as duplicated.
+     *
+     * <p>
+     * Ex) "/{x}/{y}/{z}" -> "/{}/{}/{}"
+     * </p>
+     * */
+    private final String skeleton;
+
+    /**
+     * Set of the variables declared in this dynamic path.
+     * */
+    private final Set<String> variables = Sets.newHashSet();
+
+    DynamicPath(String str) {
+        StringJoiner patternJoiner = new StringJoiner("/");
+        StringJoiner skeletonJoiner = new StringJoiner("/");
+        for (String token : str.split("/")) {
+            if (token.startsWith("{") && token.endsWith("}")) {
+                String variable = token.substring(1, token.length() - 1);
+
+                if (!variables.contains(variable)) {
+                    /*
+                     * If given token is a first occurred variable mark, add it to the variable set and replace
+                     * it to named group expression in regex.
+                     * */
+                    variables.add(variable);
+                    patternJoiner.add(String.format("(?<%s>[^/]+)", variable));
+                } else {
+                    /*
+                     * If given token is a re-occurred variable mark, replace it to backreference expression
+                     * in regex.
+                     * */
+                    patternJoiner.add(String.format("\\k<%s>", variable));
+                }
+                skeletonJoiner.add("{}");
+            } else {
+                /*
+                 * if given token is a constant, does not manipulate it.
+                 * */
+                patternJoiner.add(token);
+                skeletonJoiner.add(token);
+            }
+        }
+
+        pattern = Pattern.compile(patternJoiner.toString());
+        skeleton = skeletonJoiner.toString();
+    }
+
+    /**
+     * Returns the skeleton.
+     */
+    public String skeleton() {
+        return skeleton;
+    }
+
+    /**
+     * Returns the variables.
+     */
+    public Set<String> variables() {
+        return variables;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        DynamicPath that = (DynamicPath) o;
+        return Objects.equals(skeleton, that.skeleton);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(skeleton);
+    }
+
+    /**
+     * Check if given path matches.
+     * */
+    public boolean matches(String path) {
+        return pattern.matcher(path).matches();
+    }
+
+    /**
+     * Returns Binding results with given path.
+     * */
+    public Map<String, String> bind(String path) {
```

Comment:
Instead of having `matches()` and `bind()`, how about merging these into a single method for efficiency? If not matches, the method could return `null`.